src/RandMsg.chpl

```diff
@@ -296,9 +297,82 @@ module RandMsg
         randLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),repMsg);
         return new MsgTuple(repMsg, MsgType.NORMAL);
     }
-    
+
+    proc permutationMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+        const pn = Reflection.getRoutineName();
+        var rname = st.nextName();
+        const name = msgArgs.getValueOf("name");
+        const xName = msgArgs.getValueOf("x");
+        const size = msgArgs.get("size").getIntValue();
+        const dtypeStr = msgArgs.getValueOf("dtype");
+        const dtype = str2dtype(dtypeStr);
+        const state = msgArgs.get("state").getIntValue();
+        const isDomPerm = msgArgs.get("isDomPerm").getBoolValue();
+
+        randLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                                "name: %? size %i dtype: %? state %i isDomPerm %?".doFormat(name, size, dtypeStr, state, isDomPerm));
+
+        st.checkTable(name);
+
+        proc permuteHelper(type t, st: borrowed SymTab) throws {
+            // we need the int generator in order for permute(domain) to work correctly
+            var intGeneratorEntry: borrowed GeneratorSymEntry(int) = toGeneratorSymEntry(st.lookup(name), int);
+            ref intRng = intGeneratorEntry.generator;
+
+            if state != 1 {
+                // you have to skip to one before where you want to be
+                intRng.skipTo(state-1);
+            }
+            const permutedDom = makeDistDom(size);
+            const permutedIdx = intRng.permute(permutedDom);
+
+            if isDomPerm {
+                const permutedEntry = createSymEntry(permutedIdx);
+                st.addEntry(rname, permutedEntry);
+            }
+            else {
+                // permute requires idxType must be coercible from this stream’s eltType, so
```

Comment:
tbh I also have struggle understanding this bit. It sounds pretty clunky and strange because I normally think of "coercible to" not "coercible from".  I was gonna try and reword it but was afraid I would mess up and write something that wasn't true. So I ended up copying it verbatim from the chpl docs:
https://chapel-lang.org/docs/modules/standard/Random.html#Random.randomStream.permute

note:
The rest of this comment is my speculation / attempt at understanding... so I'm tagging @jeremiah-corrado to fact check me if he doesn't mind 😁 

I *think* what it's saying is we need to be able coerce the element type of the generator into the index type of the domain... This seems to match up with the `where` statement part of the function declaration

```chapel
proc ref permute(const ref arr: [?d] ?t): [] t  where is1DRectangularDomain(d) && isCoercible(this.eltType, d.idxType)
```

So for things like `float64` and `uint64` arrays, we use the `float64` and `uint64` generators respecitively. But our pdarrays are always indexed by `int64`s (that's the index type of the domain). We can't cast `float64` or `uint64` to an `int64` safely (we could lose information), so the where clause is false and chpl won't call this method

I tried to use and `int64` generator for all the different array types, but wasn't able to get it working. This was pretty easy to workaround in this case, but the same problem is causing me some grief with the in place `shuffle` method. I have it working for `int64` arrays but the `isCoercible` is tripping me up for the other types... I'm pretty sure I can get something working where I do some array copying, but it'd be so nice to have it just work in place on all the `symentry.a` arrays. I was planning on pinging jeremiah when I came back to this, but since he's hopefully already reading this... maybe he has a magic idea that will fix all my problems 😅

Reply:
Hi Tess, your characterization of that comment from the docs is correct. Sorry for the poor wording, it should be phrased more like: "permute requires that the stream's `eltType` is coercible to the array/domain's `idxType`"

> the same problem is causing me some grief with the in place shuffle method. I have it working for int64 arrays but the isCoercible is tripping me up for the other types

Like permute, `shuffle` needs a randomStream with an integer `eltType` to work. In this case, it looks like you're always using an `int64` randomStream. I'm probably missing something, but will the same thing work for `shuffle`?

Reply:
okayyy I could've swore I tried this! But apparently I had something else wrong because it seems to be working now!! Thanks @jeremiah-corrado :)

So a PR with `shuffle` will follow shortly